src/v/strings/static_str.h

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include <fmt/format.h>
+
+#include <string_view>
+#include <type_traits>
+
+#pragma once
+
+/**
+ * @brief A wrapper around a compile-time constant string.
+ *
+ * A static_str wraps an compile-time constant string, and is intended to be
+ * used at runtime as a guarantee that a string has static lifetime. Such string
+ * have trivial and cheap copy operations, since we just copy pointers (or
+ * views) to the underlying string.
+ */
+class static_str {
+public:
+    /**
+     * @brief Create a static_str based on a compile-time literal.
+     *
+     * The use of consteval prevents passing anything other than compile-time
+     * values.
+     */
+    consteval static_str(const char* str) // NOLINT
+      : _data{str} {
+        // do a compile-time check that the string is zero-terminated
+        std::string_view{str};
+    }
+
+    constexpr bool operator==(const static_str& rhs) const {
```

Comment:
TIL

Reply:
Yes, it is annoying. 

The reasoning is reasonable though: if you implement <=>, the compiler _could_ synthesize == from it, but it is usually less efficient, sometimes considerably so, because the code for <=> has to perform a more complicated operation, which is wasted work when you use that to implement "boolean" ==. This problem doesn't apply when everything is defaulting and the compiler is synthesizing everything.

Reply:
> compiler could synthesize == from it, but it is usually less efficient, sometimes considerably so, because the code for <=> has to perform a more complicated

my follow up question is then, if `<=>` is defaulted, why doesn't the standard say that an equivalent of `==` defaulted will also be generated.